File: model/src/main/scala/ru/mipt/acsl/decode/model/proxy/NativeLiteralGenericTypesProxyResolver.scala
package ru.mipt.acsl.decode.model.proxy

import java.util
import java.util.Optional

import ru.mipt.acsl.decode.model.Referenceable
import ru.mipt.acsl.decode.model.naming.Fqn
import ru.mipt.acsl.decode.model.proxy.path.{GenericTypeName, ProxyPath}
import ru.mipt.acsl.decode.model.registry.Registry
import ru.mipt.acsl.decode.model.types.{Alias, Const, GenericTypeSpecialized}

import scala.collection.JavaConversions._

/**
  * Created by metadeus on 20.02.16.
  */
case object NativeLiteralGenericTypesProxyResolver extends DecodeProxyResolver {

  override def resolveElement(registry: Registry, path: ProxyPath): ResolvingResult[Referenceable] = {
    path match {
      case fqnElement: ProxyPath.FqnElement =>
        resolveFqnElement(registry, fqnElement)
      case l: ProxyPath.Literal =>
        resolveLiteralElement(registry, l)
    }
  }

  private def resolveLiteralElement(registry: Registry, literal: ProxyPath.Literal): ResolvingResult[Referenceable] = {
    val ns = registry.findNamespace(Fqn.DECODE_NAMESPACE).getOrElse(sys.error("decode namespace not found"))
    val alias = new Alias.NsConst(literal.mangledName, util.Collections.emptyMap(), ns, null)
    alias.obj(Const.newInstance(alias, ns, literal.value, util.Collections.emptyList()))
    ns.objects().add(alias)
    ns.objects().add(alias.obj)
    ResolvingResult(Some(alias.obj))
  }

  private def resolveFqnElement(registry: Registry, fqnElement: ProxyPath.FqnElement): ResolvingResult[Referenceable] = {

    assert(Fqn.DECODE_NAMESPACE.size == 1, "not implemented")

    val nsFqn = fqnElement.ns
    if (nsFqn != Fqn.DECODE_NAMESPACE)
      return ResolvingResult(None)

    val systemNamespace = registry.findNamespace(Fqn.DECODE_NAMESPACE).getOrElse(sys.error("system namespace not found"))
    fqnElement.element match {
      // Generic type
      case e: GenericTypeName =>
        val maybeProxy = MaybeProxy.Type(Left(Proxy(ProxyPath(nsFqn, e.typeName))))
        val result = maybeProxy.resolve(registry)
        if (result.hasError)
          return ResolvingResult(None, result)
        val genericType = maybeProxy.obj
        val name = fqnElement.element.mangledName
        val specializedType = Option(systemNamespace.alias(name).orElse(null)).flatMap(_.obj match {
          case g: GenericTypeSpecialized => Some(g)
        }).getOrElse {
          val alias = new Alias.NsType(name, util.Collections.emptyMap(), genericType.namespace, null)
          val specializedType = GenericTypeSpecialized(alias, genericType.namespace,
            MaybeProxy.Type(Right(genericType)),
            e.genericArgumentPaths.map(arg => MaybeProxy.Type(Left(Proxy(arg)))), util.Collections.emptyList())
          alias.obj(specializedType)
          systemNamespace.objects ++= Seq(alias, specializedType)
          specializedType
        }
        val argsResult = specializedType.genericTypeArgumentsProxy.map(_.resolve(registry))
          .foldLeft(Result.empty)(_ ++ _)
        if (argsResult.hasError)
          ResolvingResult(None, argsResult)
        else
          ResolvingResult(Some(specializedType), Result.empty)
      case _ =>
        ResolvingResult(None)
    }
  }
}

File: model/src/main/java/ru/mipt/acsl/decode/model/proxy/NativeLiteralGenericTypesProxyResolver.java
package ru.mipt.acsl.decode.model.proxy;

import ru.mipt.acsl.decode.model.Referenceable;
import ru.mipt.acsl.decode.model.naming.ElementName;
import ru.mipt.acsl.decode.model.naming.Fqn;
import ru.mipt.acsl.decode.model.naming.Namespace;
import ru.mipt.acsl.decode.model.proxy.path.GenericTypeName;
import ru.mipt.acsl.decode.model.proxy.path.ProxyElementName;
import ru.mipt.acsl.decode.model.proxy.path.ProxyPath;
import ru.mipt.acsl.decode.model.proxy.path.ProxyPath$;
import ru.mipt.acsl.decode.model.registry.Registry;
import ru.mipt.acsl.decode.model.types.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * @author Artem Shein
 */
public class NativeLiteralGenericTypesProxyResolver implements DecodeProxyResolver {

    public static NativeLiteralGenericTypesProxyResolver newInstance() {
        return new NativeLiteralGenericTypesProxyResolver();
    }

    @Override
    public ResolvingResult<Referenceable> resolveElement(Registry registry, ProxyPath path) {
        if (path instanceof ProxyPath.FqnElement)
            return resolveFqnElement(registry, (ProxyPath.FqnElement) path);
        else if (path instanceof ProxyPath.Literal)
            return resolveLiteralElement(registry, (ProxyPath.Literal) path);
        else
            throw new RuntimeException("not implemented");
    }

    private ResolvingResult<Referenceable> resolveLiteralElement(Registry registry, ProxyPath.Literal literal) {
        Namespace ns = registry.findNamespace(Fqn.DECODE_NAMESPACE)
                .orElseThrow(() -> new RuntimeException("decode namespace not found"));
        Alias.NsConst alias = new Alias.NsConst(literal.mangledName(), new HashMap<>(), ns, null);
        alias.obj(Const.newInstance(alias, ns, literal.value(), new ArrayList<>()));
        ns.objects().add(alias);
        ns.objects().add(alias.obj());
        return ResolvingResult.newInstance(alias.obj());
    }

    private ResolvingResult<Referenceable> resolveFqnElement(Registry registry, ProxyPath.FqnElement fqnElement) {

        if(Fqn.DECODE_NAMESPACE.size() != 1)
            throw new AssertionError("not implemented");

        Fqn nsFqn = fqnElement.ns();
        if (!nsFqn.equals(Fqn.DECODE_NAMESPACE))
            return ResolvingResult.newInstance();

        Namespace systemNamespace = registry.findNamespace(Fqn.DECODE_NAMESPACE)
                .orElseThrow(() -> new AssertionError("system namespace not found"));
        ProxyElementName element = fqnElement.element();
        if (!(element instanceof GenericTypeName))
            return ResolvingResult.newInstance();
        GenericTypeName e = (GenericTypeName) element;
        MaybeTypeProxyType maybeProxy = MaybeTypeProxyType.newInstance(Proxy$.MODULE$.apply(
                ProxyPath$.MODULE$.apply(nsFqn, e.typeName())));
        ResolvingMessages result = maybeProxy.resolve(registry);
        if (result.hasError())
            return ResolvingResult.newInstance(null, result);
        DecodeType genericType = maybeProxy.obj();
        ElementName name = element.mangledName();
        GenericTypeSpecialized specializedType = systemNamespace.alias(name).flatMap(a -> a.obj() instanceof GenericTypeSpecialized ?
                Optional.of((GenericTypeSpecialized) a.obj()) : Optional.empty()).orElseGet(() -> {
            Alias.NsType alias = new Alias.NsType(name, new HashMap<>(), genericType.namespace(), null);
            GenericTypeSpecialized newSpecializedType =
                    GenericTypeSpecialized$.MODULE$.apply(alias, genericType.namespace(),
                            MaybeTypeProxyType.newInstance(genericType),
                            e.genericArgumentPaths().stream()
                                    .map(arg -> MaybeTypeProxyType.newInstance(Proxy$.MODULE$.apply(arg)))
                                    .collect(Collectors.toList()),
                            new ArrayList<>());
            alias.obj(newSpecializedType);
            List<Referenceable> objects = systemNamespace.objects();
            objects.add(alias);
            objects.add(newSpecializedType);
            return newSpecializedType;
        });
        ResolvingMessages argsResult = specializedType.genericTypeArgumentsProxy().stream().map(p -> p.resolve(registry))
                .reduce(ResolvingMessages.newInstance(), (l, r) -> { l.addAll(r); return l; });
        return argsResult.hasError()
                ? ResolvingResult.newInstance(null, argsResult)
                : ResolvingResult.newInstance(specializedType);
    }

    private NativeLiteralGenericTypesProxyResolver() {

    }
}

